### ISSUE TITLE
[no-unused-vars] TypeError: context.report() called with a messageId, but no messages were present in the rule metadata.

### ISSUE THREAD

**Issue by esetnik** [typescript-eslint/typescript-eslint#2451]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```js
module.exports = {
  extends: [
    'react-app',
    'eslint:recommended',
    'plugin:@typescript-eslint/eslint-recommended',
    'plugin:@typescript-eslint/recommended',
    'plugin:prettier/recommended',
  ],
  rules: {
    '@typescript-eslint/no-unused-vars': [1, { argsIgnorePattern: '^_' }],
    '@typescript-eslint/explicit-function-return-type': 'off',
    'no-restricted-imports': [
      'error',
      {
        paths: [
          {
            name: 'styled-components',
            message: 'Please import from styled-components/macro.',
          },
        ],
        patterns: ['!styled-components/macro'],
      },
    ],
  },
};

```

```TS
// jest-styled-components.d.ts
declare namespace jest {
  interface Matchers<R> {
    toHaveStyleRule: import('jest-styled-components').jest.Matchers<
      R
    >['toHaveStyleRule'];
  }
}
```

```json
{
  "compilerOptions": {
    "target": "es5",
    "lib": ["dom", "dom.iterable", "esnext"],
    "allowJs": true,
    "skipLibCheck": true,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "noEmit": true,
    "jsx": "react"
  },
  "include": ["src", "test"],
  "plugins": [
    {
      "name": "typescript-styled-plugin"
    }
  ]
}
```

**Expected Result**

No crash. The same configuration and file works fine in v3.10.1

**Actual Result**

After upgrading to v4.0.1
```
TypeError: context.report() called with a messageId, but no messages were present in the rule metadata.
Occurred while linting /Users/redacted/frontend/src/types/jest-styled-components.d.ts:1
    at /Users/redacted/frontend/node_modules/eslint/lib/linter/report-translator.js:317:23
    at Object.report (/Users/redacted/frontend/node_modules/eslint/lib/linter/linter.js:905:41)
    at TSModuleDeclaration (/Users/redacted/frontend/node_modules/@typescript-eslint/eslint-plugin/dist/rules/no-unused-vars.js:124:33)
    at /Users/redacted/frontend/node_modules/eslint/lib/linter/safe-emitter.js:45:58
    at Array.forEach (<anonymous>)
    at Object.emit (/Users/redacted/frontend/node_modules/eslint/lib/linter/safe-emitter.js:45:38)
    at NodeEventGenerator.applySelector (/Users/redacted/frontend/node_modules/eslint/lib/linter/node-event-generator.js:254:26)
    at NodeEventGenerator.applySelectors (/Users/redacted/frontend/node_modules/eslint/lib/linter/node-event-generator.js:283:22)
    at NodeEventGenerator.enterNode (/Users/redacted/frontend/node_modules/eslint/lib/linter/node-event-generator.js:297:14)
    at CodePathAnalyzer.enterNode (/Users/redacted/frontend/node_modules/eslint/lib/linter/code-path-analysis/code-path-analyzer.js:634:23)
```

**Additional Info**

[typescript-eslint-debug.log](https://github.com/typescript-eslint/typescript-eslint/files/5152929/typescript-eslint-debug.log)

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.0.1` |
| `@typescript-eslint/parser`        | `4.0.1` |
| `TypeScript`                       | `4.0.2` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `12.8.3` |

**Comment by bradzacher** [comment]
This happens because you're on ESLint v6, but the messageID we're using was released in v7.

**Comment by esetnik** [comment]
@bradzacher is @typescript-eslint v4.0.1 supposed to be compatible with eslint v6? I see it is included in the `peerDeps` range.

https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/package.json#L64

**Comment by bradzacher** [comment]
Which is why I marked this as a bug.
It _should_ work with v6 but it doesn't.

**Comment by esetnik** [comment]
Ok now I understand. I'm not too familiar with the project but I'm happy to work on a PR or provide any additional info if you can point me in the right direction.

**Comment by bradzacher** [comment]
It's actually a super simple fix.

https://github.com/typescript-eslint/typescript-eslint/blob/46ad4d051d8b58e003cb70f7f449218b5d354391/packages/eslint-plugin/src/rules/no-unused-vars.ts#L24

Has to handle the case where there are no messages defined, and default to something like the current v7 messages config:

https://github.com/eslint/eslint/blob/9250ab3e63836c71af87b10f29157bed278054e7/lib/rules/no-unused-vars.js#L78-L80

I.e. the above line becomes something like

```ts
messages: baseRule.meta.messages ?? {
  unusedVar: "'{{varName}}' is {{action}} but never used{{additional}}."
},
```

Happy to accept a PR if you want to do it!

**Comment by esetnik** [comment]
It looks like a lot of the other rules need to be fixed e.g. [no-unused-expressions](https://github.com/typescript-eslint/typescript-eslint/blob/46ad4d051d8b58e003cb70f7f449218b5d354391/packages/eslint-plugin/src/rules/no-unused-expressions.ts#L22). Do you want one PR to fix them all?

**Comment by bradzacher** [comment]
Yeah definitely. That'll be much easier to submit and merge. Go for it!

**Comment by esetnik** [comment]
I'm also thinking that the automated test suite should include different eslint versions in the test matrix to catch issues like this automatically. I'm looking into the best way to achieve that.

**Comment by bradzacher** [comment]
Yeah testing old versions of things has always been a problem for us.

Usually eslint isn't too much of a problem because its api and internals are pretty stable for our use case.

We have an integration test for v6, but it's mostly there to make sure that we don't do any incompatible deep requires from eslint.

If you can figure out a good way to do it, that'd be fantastic!

**Comment by esetnik** [comment]
@bradzacher I have implemented the missing messages fallbacks in #2461. I will look to improving the test configuration to support testing multiple eslint versions in a separate PR.

**Comment by esetnik** [comment]
@bradzacher I have implemented eslint version matrix in test runner in #2463. It is currently failing because it is based off master (which is not compatible with eslint v5 or v6 due to #2451). The tests should start to pass after it is rebased off #2461.

**Comment by esetnik** [comment]
@bradzacher Will you be publishing a new version including this fix in the near future? It's currently the only thing blocking us from using the new @typescript-eslint with create-react-app.

**Comment by bradzacher** [comment]
We release every Monday morning (PST).

This weekend I'm planning on fixing all of the scope analysis bugs people found in this version (see milestone), so Monday's version will be a much better version for you to upgrade CRA to.

**Comment by jarretmoses** [comment]
@bradzacher Just for clarification I am getting this error too but it only occurs when trying to lint declaration files. If I use `ignorePatterns: ['*.d.ts']` it seems to work fine. Does this resolve this issue?
